boa_gc/src/internals/eph_box.rs

```diff
@@ -0,0 +1,160 @@
+        EphemeronBox {
+            key: Cell::new(Some(key.inner_ptr())),
+            value,
+        }
+    }
+}
+
+impl<K: Trace + ?Sized, V: Trace + ?Sized> EphemeronBox<K, V> {
+    /// Checks if the key pointer is marked by Trace
+    #[inline]
+    pub(crate) fn is_marked(&self) -> bool {
+        if let Some(key) = self.inner_key() {
+            key.is_marked()
+        } else {
+            false
+        }
+    }
+
+    /// Returns some pointer to the `key`'s `GcBox` or None
+    /// # Panics
+    /// This method will panic if called while the garbage collector is dropping.
+    #[inline]
+    fn inner_key_ptr(&self) -> Option<*mut GcBox<K>> {
+        assert!(finalizer_safe());
+        self.key.get().map(NonNull::as_ptr)
+    }
+
+    /// Returns some reference to `key`'s `GcBox` or None
+    #[inline]
+    fn inner_key(&self) -> Option<&GcBox<K>> {
+        // SAFETY: This is safe as `EphemeronBox::inner_key_ptr()` will
+        // fetch either a live `GcBox` or None. The value of `key` is set
+        // to None in the case where `EphemeronBox` and `key`'s `GcBox`
+        // entered into `Collector::sweep()` as unmarked.
+        unsafe {
+            if let Some(inner_key) = self.inner_key_ptr() {
+                Some(&*inner_key)
+            } else {
+                None
+            }
+        }
+    }
+
+    /// Returns a reference to the value of `key`'s `GcBox`
+    #[inline]
+    pub(crate) fn key(&self) -> Option<&K> {
+        if let Some(key_box) = self.inner_key() {
+            Some(key_box.value())
+        } else {
+            None
+        }
+    }
+
+    /// Returns a reference to `value`
+    #[inline]
+    pub(crate) fn value(&self) -> &V {
+        &self.value
+    }
+
+    /// Calls [`Trace::weak_trace()`][crate::Trace] on key
+    #[inline]
+    fn weak_trace_key(&self) {
+        if let Some(key) = self.inner_key() {
+            key.weak_trace_inner();
+        }
+    }
+
+    /// Calls [`Trace::weak_trace()`][crate::Trace] on value
+    ///
+    #[inline]
+    fn weak_trace_value(&self) {
+        // SAFETY: Value is a sized element that must implement trace. The
+        // operation is safe as EphemeronBox owns value and `Trace::weak_trace`
+        // must be implemented on it
+        unsafe {
+            self.value().weak_trace();
+        }
+    }
+}
+
+// `EphemeronBox`'s Finalize is special in that if it is determined to be unreachable
+// and therefore so has the `GcBox` that `key`stores the pointer to, then we set `key`
+// to None to guarantee that we do not access freed memory.
+impl<K: Trace + ?Sized, V: Trace + ?Sized> Finalize for EphemeronBox<K, V> {
+    #[inline]
+    fn finalize(&self) {
+        self.key.set(None);
+    }
+}
+
+// SAFETY: Please see [`Trace]
+unsafe impl<K: Trace + ?Sized, V: Trace + ?Sized> Trace for EphemeronBox<K, V> {
+    /// # Safety
+    ///
+    /// Please see [`Trace`].
+    #[inline]
+    unsafe fn trace(&self) {
```

Comment:
I'm pretty sure implementations of unsafe methods of traits don't need a `Safety` section. That section is provided by the trait definition. Also applies to the added documentation.